keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,1330 @@
+The pods would be selected according to the node (`.spec.nodeSelector`) and a subset of the default
+kubectl [drain filters](https://github.com/kubernetes/kubernetes/blob/56cc5e77a10ba156694309d9b6159d4cd42598e1/staging/src/k8s.io/kubectl/pkg/drain/filters.go#L153-L162).
+
+Used drain filters:
+- `daemonSetFilter`, skips daemon sets to keep critical workloads alive.
+- `mirrorPodFilter`, skips static mirror pods.
+
+Omitted drain filters:
+- `skipDeletedFilter`: updating the condition of already terminating pods should have no
+  downside and will be informative for the user.
+- `unreplicatedFilter`: actors who own pods without a controller owner reference should have the
+  opportunity to evacuate their pods. It is a noop if the owner does not respond.
+- `localStorageFilter`, we can leave the responsibility of whether to evacuate a pod with  local
+  storage (having `EmptyDir` volumes) to the owning workload. For example, a controller of a
+  deployment that has a `.spec.strategy.rollingUpdate.maxSurge` defined assumes that it is safe to
+  remove the pod and the `EmptyDir` volume.
+
+The selection process can be later enhanced to target daemon set pods according to the priority or
+pod type.
+
+Controllers that own these marked pods, would observe them and start a removal or migration from
+the nodes upon detecting the `EvacuationRequested` condition. They will also indicate this by
+setting the `EvacuationInitiated` condition on the pod.
+
+The node maintenance controller would also remove the `EvacuationRequested` condition from the
+targeted pods if the `.spec.stage` of the NodeMaintenance is set to `Complete`, unless there is no
+other maintenance in progress. The condition will only be removed if the reason of the condition
+is `NodeMaintenance$NAME`, where NAME is the name of the NodeMaintenance. If the reason has a
+different value, then it is owned by another controller (e.g. descheduler) or by a different
+NodeMaintenance object and we should keep the condition.
+
+The controller can show progress by reconciling:
+- `.status.nodes[0].PodsPendingEvacuation`, to show how many pods remain to be removed from
+  the first node.
+- `.status.nodes[0].PodsEvacuating`, to show how many pods have been accepted for the
+  evacuation from the first node. These are the pods that have the `EvacuationInitiated`
+  condition set to `True`.
+- To keep track of the entire maintenance the controller will reconcile a `Drained` condition and
+  set it to true if all pods pending evacuation/termination have terminated on all target nodes
+  when drain is requested by the maintenance object.
+- NodeMaintenance condition or annotation can be set on the node object to advertise the current
+  phase of the maintenance.
+
+#### Supported Stage Transitions
+
+The following transitions should be validated by the API server.
+
+- Idle -> _Deletion_
+  - Planning a maintenance in the future and canceling/deleting it without any consequence.
```

Comment:
Tracking planned maintenance allows an autoscaling controller to create a new node ahead of the maintenance starting, preserving cluster capacity.

It's a capacity surge model rather than a repair model.

Reply:
I see the value of the idle stage, but what would initiate moving from _Idle_ to _Cordon_. I didn't understand how and when it will change.

Reply:
It is an API, so either manually by the cluster admin or by a controller. Note that the user can create the NodeMaintenance with `stage: Drain` directly. This is just provides more options.